### PR TITLE
BUG: Index.drop raising Error when Index has duplicates

### LINKED ISSUE #33494: BUG: MultiIndex drop produces wrong output if not lexicographically sorted
- [x] I have checked that this issue has not already been reported.

- [x] I have confirmed this bug exists on the latest version of pandas.
  Latest Pandas version my system installs (via `anaconda update pandas`), which is 0.25.3.

- [ ] (optional) I have confirmed this bug exists on the master branch of pandas.

---

#### Code Sample, a copy-pastable example

```python
import pandas as pd

df1 = pd.DataFrame(
    [
        [4, 5, 6],
        [5, 4, 6],
        [4, 5, 6],
        [3, 7, 5],
        [5, 9, 0],
        [1, 2, 3],
    ],
    columns=["a", "b", "c"],
)

df2 = pd.DataFrame(
    [
        [1, 2, 3],
        [4, 5, 6],
        [1, 6, 0],
    ],
    columns=["a", "b", "c"],
)

df1mi = pd.MultiIndex.from_frame(df1)
df2mi = pd.MultiIndex.from_frame(df2)
ismi = df1mi.intersection(df2mi)  # intersection of df1 and df2
# df1mi, _ = df1mi.sortlevel(sort_remaining=True)
complement = df1mi.drop(ismi).to_frame().reset_index(drop=True)

print(df1, "\n")
print(df2, "\n")
print(complement)
```

#### Problem description

Above code computes the multi-set complement A-B for two multi-sets A and B (only difference to sets is that duplicates in A that are not in B are preserved).

Without the commented-out line, pandas issues a warning (`sys:1: PerformanceWarning: indexing past lexsort depth may impact performance.`). However, the result is wrong: Row 1 from df1 is missing in the complement.

With the commented-out line, no warning is issued and the result is correct.

The bug is that without the sorting, the drop command malfunctions. 

#### Expected Output

Correct output (with explicit sort):

```
   a  b  c
0  3  7  5
1  5  4  6
2  5  9  0
```

Wrong output (without explicit sort):

```
   a  b  c
0  3  7  5
1  5  9  0
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit           : None
python           : 3.7.7.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 18.7.0
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : en_US.UTF-8
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 0.25.3
numpy            : 1.18.2
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 20.0.2
setuptools       : 46.1.3.post20200330
Cython           : None
pytest           : 5.4.1
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.11.1
IPython          : 7.13.0
pandas_datareader: None
bs4              : 4.8.2
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : 3.2.1
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
s3fs             : None
scipy            : 1.4.1
sqlalchemy       : None
tables           : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None

</details>

### LINKED ISSUE #38051: BUG: Index.drop raises on non-unique Index
```
index = pd.Index(range(3)).repeat(2)

>>> index.drop(1)
Traceback (most recent call last):
[...]
pandas.errors.InvalidIndexError: Reindexing only valid with uniquely valued Index objects
```

It is using Index.get_indexer, which is only for uniques indexes.

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #38051
- [x] closes #33494
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

@jbrockmendel  This should deal with duplicates. For MultiIndex sometimes a slice with stepzize greater than zero was given, which dropped to many elements

### CODE REVIEW

File: pandas/core/indexes/base.py [hunk 1]
```diff
@@ -5508,7 +5508,10 @@ def drop(self, labels, errors: str_t = "raise"):
         """
         arr_dtype = "object" if self.dtype == "object" else None
         labels = com.index_labels_to_array(labels, dtype=arr_dtype)
-        indexer = self.get_indexer(labels)
+        if self.is_unique:
+            indexer = self.get_indexer(labels)
+        else:
+            indexer, _ = self.get_indexer_non_unique(labels)
```
Comment (jbrockmendel) [inline pandas/core/indexes/base.py:5514]: get_indexer_for?

Comment (phofl) [reply pandas/core/indexes/base.py:5514]: Did not know this function, thanks very much.

File: pandas/tests/indexes/test_base.py [hunk 2]
```diff
@@ -1496,6 +1496,13 @@ def test_drop_tuple(self, values, to_drop):
             with pytest.raises(KeyError, match=msg):
                 removed.drop(drop_me)
 
+    def test_drop_with_duplicates_in_index(self):
```
Comment (jbrockmendel) [inline pandas/tests/indexes/test_base.py:1499]: you could use the `index` fixture and do something like

```
def test_drop_with_duplicates(self, index):
    if len(index) == 0: return
    index = index.repeat(2)  #  <-- ensure duplicates
    res = index.drop(index[0])
```

Comment (phofl) [reply pandas/tests/indexes/test_base.py]: Better? Is there a way to construc expected without drop?

File: pandas/tests/indexes/test_base.py [hunk 3]
```diff
@@ -1494,6 +1494,15 @@ def test_drop_tuple(self, values, to_drop):
             with pytest.raises(KeyError, match=msg):
                 removed.drop(drop_me)
 
+    def test_drop_with_duplicates_in_index(self, index):
+        # GH38051
+        if len(index) == 0:
+            return
+        expected = index.drop(index[0]).repeat(2)
```
Comment (jbrockmendel) [inline pandas/tests/indexes/test_base.py:1501]: this works, but ideally we'd form `expected` without using `drop`.  could do

```
index = index.unique()
index = index.repeat(2)
expected = index[2:]
result = index.drop(index[0])
```

Comment (phofl) [reply pandas/tests/indexes/test_base.py:1501]: Had a similar idea, but the fixture contains indexes with duplicates. If drop is bad, we could use unique and index[1:] atfterwards?

Comment (jbrockmendel) [reply pandas/tests/indexes/test_base.py:1501]: can you write your idea out explicitly? im not clear on how it is different from what i wrote

**Comment by phofl** [comment]
Exited for the MultiIndex cases, because would have to catch a PerformanceWarning there if it is not sorted and we are testing this at another place too.

File: pandas/tests/indexes/test_base.py (hunk 3 above)
Comment (phofl) [reply pandas/tests/indexes/test_base.py:1501]: Sorry, that does not work either. Example to illustrate the problem:
```
index = Index([1, 2, 1, 2])
```
index[1:] is then Index([2, 1, 2])
Repeating and dropping one in the index leads to
```
result = Index([2,2,2,2])
```
while index[1:].repeat(2) would be
```
expected = Index([2,1,2,2,1,2])
```

**Comment by jreback** [comment]
looks good, @phofl can you merge master and ping on green.

File: pandas/tests/indexes/multi/test_drop.py [hunk 4]
```diff
@@ -149,6 +151,16 @@ def test_drop_with_nan_in_index(nulls_fixture):
         mi.drop(pd.Timestamp("2001"), level="date")
 
 
+def test_drop_with_non_monotonic_duplicates():
+    # GH#33494
+    mi = MultiIndex.from_tuples([(1, 2), (2, 3), (1, 2)])
+    with warnings.catch_warnings():
```
Comment (simonjayhawkins) [inline pandas/tests/indexes/multi/test_drop.py:157]: could maybe use `@pytest.mark.filterwarnings` as a follow-up

**Comment by phofl** [comment]
Thanks for merging. Found that warnings call in another Test. Will look through as a follow up and clean this up

**Comment by simonjayhawkins** [comment]
@phofl test_drop_with_duplicates_in_index failing on 32bit (not all tests completed on other envs yet)

**Comment by phofl** [comment]
Dtype casting issue, will look into this later.

**Comment by phofl** [comment]
I think this is a configuration which can not work on 32 bit for this one test, so I skipped it

**Comment by phofl** [comment]
@jreback green

**Comment by jreback** [comment]
thanks @phofl

**Pull request merged**